File: Data Structures & Algorithms/find-minimum-in-rotated-sorted-array/submission-4.py
class Solution:
    def findMin(self, nums: List[int]) -> int:
        if len(nums) == 1:
            return nums[0]

        l, r = 0, len(nums) - 1

        if nums[l] < nums[r]:
            return nums[l]
        
        new = [0] * len(nums)
        new[0] = nums[-1]
        for i in range(1, len(nums)):
            new[i] = nums[i - 1]

        return self.findMin(new)
        

File: Data Structures & Algorithms/combination-target-sum/submission-4.java
class Solution {
    private void backtrack(int[] nums, int target, int i, int currentTotal, List<Integer> subset, List<List<Integer>> res) {
        if (currentTotal == target) {
            res.add(new ArrayList<>(subset));
            return;
        }
        if (i >= nums.length) {
            return;
        }

        // include the current num as much as we can
        if (currentTotal + nums[i] <= target) {
            subset.add(nums[i]);
            backtrack(nums, target, i, currentTotal + nums[i], subset, res);
            subset.remove(subset.size() - 1);
        }

        // skip this number
        backtrack(nums, target, i + 1, currentTotal, subset, res);

    }

    public List<List<Integer>> combinationSum(int[] nums, int target) {
        List<List<Integer>> res = new ArrayList<>();
        List<Integer> subset = new ArrayList<>();

        backtrack(nums, target, 0, 0, subset, res);
        return res;
    }

}
